packages/vite/src/node/plugins/resolve.ts

```diff
@@ -636,8 +635,17 @@ export function tryNodeResolve(
       return resolved
     }
     const resolvedExt = path.extname(resolved.id)
-    const resolvedId =
-      isDeepImport && path.extname(id) !== resolvedExt ? id + resolvedExt : id
+    let resolvedId = id
+    if (isDeepImport) {
+      // check ext before externalizing - only externalize
+      // extension-less imports and explicit .js imports
+      if (resolvedExt && !resolved.id.match(/(.js|.mjs|.cjs)$/)) {
+        return
```

Comment:
Is this meant to be `return resolved`?

Reply:
No, it should return undefined. The only way this code is reached is when called to test if it should be externalized here https://github.com/vitejs/vite/blob/e6f3b026bd1e281cfad6b19386195b6c2dc94165/packages/vite/src/node/ssr/ssrExternal.ts#L168
Once this returns undefined, the second tryNodeResolve in the resolve plugin will already have `externalize` false so it wont reach the branch. This is quite confusing though. It was done as part of a refactoring that tried to change too much of the current structure. Now that we know the new SSR externals logic is working, we should do another refactoring to clean up and avoid the need to calling tryNodeResolve twice (the ssr check is cached, so it is only duplicated the first time but still)

Reply:
Oh wow that is very confusing, haha 😆